x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/optimizer/TomlFoldTests.java

```diff
@@ -0,0 +1,118 @@
+        List<EqlFoldSpec> foldSpecs = EqlFoldSpecLoader.load("/test_folding.toml");
+        foldSpecs.addAll(EqlFoldSpecLoader.load("/test_string_functions.toml"));
+        List<EqlFoldSpec> unsupportedSpecs = EqlFoldSpecLoader.load("/test_unsupported.toml");
+
+        HashSet<EqlFoldSpec> filteredSpecs = new HashSet<>(foldSpecs);
+        filteredSpecs.removeAll(unsupportedSpecs);
+        return asArray(filteredSpecs);
+    }
+
+    public static List<Object[]> asArray(Collection<EqlFoldSpec> specs) {
+        AtomicInteger counter = new AtomicInteger();
+        return specs.stream().map(spec -> new Object[] {
+            counter.incrementAndGet(), spec
+        }).collect(toList());
+    }
+
+    private static final EqlConfiguration caseSensitiveConfig = config(true);
+    private static final EqlConfiguration caseInsensitiveConfig = config(false);
+    private static EqlParser parser = new EqlParser();
+    private static final EqlFunctionRegistry functionRegistry = new EqlFunctionRegistry();
+
+    private final int num;
+    private final EqlFoldSpec spec;
+
+    private static EqlConfiguration config(boolean isCaseSensitive) {
+        return new EqlConfiguration(new String[]{"none"},
+            org.elasticsearch.xpack.ql.util.DateUtils.UTC, "nobody", "cluster",
+            null, TimeValue.timeValueSeconds(30), -1, false, isCaseSensitive, "",
+            new TaskId("foobarbaz", 1234567890), () -> false);
+    }
+
+    public TomlFoldTests(int num, EqlFoldSpec spec) {
+        this.num = num;
+        this.spec = spec;
+    }
+
+    private void testWithConfig(EqlConfiguration testConfig, EqlFoldSpec spec) {
+        Expression expr = parser.createExpression(spec.expression());
+
+        // resolve all function calls
+        expr = expr.transformUp(e -> {
+            if (e instanceof UnresolvedFunction) {
+                UnresolvedFunction uf = (UnresolvedFunction) e;
+
+                if (uf.analyzed()) {
+                    return uf;
+                }
+
+                String name = uf.name();
+
+                if (uf.childrenResolved() == false) {
+                    return uf;
+                }
+
+                String functionName = functionRegistry.resolveAlias(name);
+                if (functionRegistry.functionExists(functionName) == false) {
+                    return uf.missing(functionName, functionRegistry.listFunctions());
+                }
+                FunctionDefinition def = functionRegistry.resolveFunction(functionName);
+                Function f = uf.buildResolved(testConfig, def);
+                return f;
+            }
+            return e;
+        });
+
+        assertTrue(expr.foldable());
+        Object folded = expr.fold();
+
+        // upgrade to a long, because the parser typically downgrades Long -> Integer when possible
+        if (folded instanceof Integer) {
+            folded  = ((Integer) folded).longValue();
+        }
```

Comment:
Can you expand on this one? Which parser does the downgrading and why is that a bad thing? Is it us or or the TOML spec that doesn't preserve the type?

Reply:
TOML will give us Long or Double.

This is where the downgrading is happening in the parser:
https://github.com/elastic/elasticsearch/blob/3e39e8097c2f54f2b3f860b213cc672710b03d50/x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/ExpressionBuilder.java#L211-L215